File: Day75/Part6/endless_wave.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int n, K;
    long long S;
    if (cin >> n >> K >> S) {
        if (n == 0) {
            cout << "1" << endl;
            return 0;
        }
        vector<long long> p(n);
        for (int i = 0; i < n; ++i) cin >> p[i];

        long long active_power = 0;
        bool possible = true;
        
        for (int i = 0; i < n; ++i) {
            active_power += p[i];
            
            // Can defeat K enemies per unit time
            // So we can reduce active_power by K (defeating strongest or any K)
            long long defeated = min((long long)K, active_power);
            active_power -= defeated;
            
            if (active_power > S) {
                possible = false;
                break;
            }
        }

        cout << (possible ? "1" : "0") << endl;
    }
    return 0;
}

File: Day75/Part6/endless_wave.py
def solve():
    try:
        line = input().split()
        if not line:
            return
        
        n, K, S = int(line[0]), int(line[1]), int(line[2])
        
        if n == 0:
            print("1")
            return
            
        line2 = input().split()
        p = [int(x) for x in line2]
        
        active_power = 0
        possible = True
        
        for power in p:
            active_power += power
            
            defeated = min(K, active_power)
            active_power -= defeated
            
            if active_power > S:
                possible = False
                break
                
        print("1" if possible else "0")
            
    except EOFError:
        pass

if __name__ == "__main__":
    solve()
